### ISSUE TITLE
Support for named parameters

### ISSUE THREAD

**Issue by cytrinox** [fernandobatels/rsfbclient#48]
Please consider to include support for namend SQL parameters:

`SELECT * FROM test WHERE customerId = :customer_id`

Firebird supports this and it is much safer than using unnamed question marks.

**Comment by fernandobatels** [comment]
You have a have an utilization example of this approach in rust?

**Comment by cytrinox** [comment]
I like the way rusqlite has solved it:


````rust
struct Person {
    name: String,
    age_in_years: u8,
    data: Option<Vec<u8>>,
}

fn add_person(conn: &Connection, person: &Person) -> Result<()> {
    conn.execute_named(
        "INSERT INTO person (name, age_in_years, data)
         VALUES (:name, :age, :data)",
        named_params!{
            ":name": person.name,
            ":age": person.age_in_years,
            ":data": person.data,
        }
    )?;
    Ok(())
}
````

See https://docs.rs/rusqlite/0.24.0/rusqlite/macro.named_params.html

**Comment by fernandobatels** [comment]
Very interesting. I think that we can implement this in the future. You have interest in help us implementing this?

**Comment by fernandobatels** [comment]
Perhaps it would be more interesting if we can use the struct directly instead of this 'hash map' approach.
Something like this:

```rust
struct Person {
    name: String,
    age_in_years: u8,
    data: Option<Vec<u8>>,
}

fn add_person(conn: &Connection, person: &Person) -> Result<()> {
    conn.execute_named(
        "INSERT INTO person (name, age_in_years, data)
         VALUES (:name, :age, :data)",
        person
    )?;

    // Or using a macro
     conn.execute_named(
        "INSERT INTO person (name, age_in_years, data)
         VALUES (:name, :age, :data)",
         to_named_params!(person)
    )?;

    Ok(())
}
```

**Comment by jairinhohw** [comment]
> Perhaps it would be more interesting if we can use the struct directly instead of this 'hash map' approach.
> Something like this:
> 
> ```rust
> struct Person {
>     name: String,
>     age_in_years: u8,
>     data: Option<Vec<u8>>,
> }
> 
> fn add_person(conn: &Connection, person: &Person) -> Result<()> {
>     conn.execute_named(
>         "INSERT INTO person (name, age_in_years, data)
>          VALUES (:name, :age, :data)",
>         person
>     )?;
> 
>     // Or using a macro
>      conn.execute_named(
>         "INSERT INTO person (name, age_in_years, data)
>          VALUES (:name, :age, :data)",
>          to_named_params!(person)
>     )?;
> 
>     Ok(())
> }
> ```

To do allow using a struct directly like this would require the struct to implement some trait, so we would need to
create a procedural macro to implement it for it to be usable, like the `Queryable` trait of diesel.
It would be easier to make it like the sqlite does, with a normal macro.

**Comment by jairinhohw** [comment]
> Please consider to include support for namend SQL parameters:
> 
> `SELECT * FROM test WHERE customerId = :customer_id`
> 
> Firebird supports this and it is much safer than using unnamed question marks.

Firebird does not seem to allow this syntax directly, but it can be emulated and converted to positional parameters
in the lib, or am i missing something?

**Comment by fernandobatels** [comment]
> > Perhaps it would be more interesting if we can use the struct directly instead of this 'hash map' approach.
> > Something like this:
> > ```rust
> > struct Person {
> >     name: String,
> >     age_in_years: u8,
> >     data: Option<Vec<u8>>,
> > }
> > 
> > fn add_person(conn: &Connection, person: &Person) -> Result<()> {
> >     conn.execute_named(
> >         "INSERT INTO person (name, age_in_years, data)
> >          VALUES (:name, :age, :data)",
> >         person
> >     )?;
> > 
> >     // Or using a macro
> >      conn.execute_named(
> >         "INSERT INTO person (name, age_in_years, data)
> >          VALUES (:name, :age, :data)",
> >          to_named_params!(person)
> >     )?;
> > 
> >     Ok(())
> > }
> > ```
> 
> To do allow using a struct directly like this would require the struct to implement some trait, so we would need to
> create a procedural macro to implement it for it to be usable, like the `Queryable` trait of diesel.
> It would be easier to make it like the sqlite does, with a normal macro.

Yes, but making the struct approach, instead of the sqlite does, we will stay more close an ORM support and will make the utilization more easy.

**Comment by fernandobatels** [comment]
> > Please consider to include support for namend SQL parameters:
> > `SELECT * FROM test WHERE customerId = :customer_id`
> > Firebird supports this and it is much safer than using unnamed question marks.
> 
> Firebird does not seem to allow this syntax directly, but it can be emulated and converted to positional parameters
> in the lib, or am i missing something?

Hmm, now I can't response this, but I think that each lib implement this own approach

**Comment by fernandobatels** [comment]
For me, any of two approach are welcome. Maybe we can implement both. But I think that we must use separated crate for this 'advanced' API's

**Comment by jairinhohw** [comment]
> For me, any of two approach are welcome. Maybe we can implement both. But I think that we must use separated crate for this 'advanced' API's

Well, we would need a separate crate to create procedural macros anyways

**Comment by juarezr** [comment]
> > For me, any of two approach are welcome. Maybe we can implement both. But I think that we must use separated crate for this 'advanced' API's
> 
> Well, we would need a separate crate to create procedural macros anyways

Maybe would be more interesting adding support for `rsfbclient` in the existing Rust ORMs that redoing the ORM inside `rsfbclient`.

Notice that [sqlx](https://docs.rs/sqlx/0.3.5/sqlx/macro.query.html) already does something very similar that you are expecting here.

Taking an approach like this could be good for both projects because one could join forces and give more features faster for the users.

: smiley:

**Comment by afsmi** [comment]
I think @jairinhohw is right and most Firebird DSQL does not actually support named parameters in the way proposed above, except maybe inside `EXECUTE STATEMENT` and `EXECUTE BLOCK` statements (but no chance for us to deal with that at compile time without parsing the statement in our library)

I think rusqlite depends on SQLite supporting named parameters directly, or maybe it does parsing of the SQL. If rusqlite is doing validation on the provided types of parameters, but doesn't parse the SQL, then likely any such validation is happening using dynamic facilities of SQLite ("Hey C implementation of Sqlite, did i provide correct types for these parameters?").

Currently, if users don't mind tweaking their SQL a little bit they could just use `format!` macro or we can provide one ourselves that wraps the result in an execute call

```
let stmt = format!("INSERT INTO person (name, age_in_years, data)
         VALUES ({name}, {age}, {data})", name="'afsmi'", age="3", data="'afsmi is 3 years old'" );
```
but obviously they cannot easily use their existing queries this way, and it does not prevent runtime errors that occur due to database type mismatches.

we could also make a format-like macro that does something like
```
let stmt = our_macro!("INSERT INTO person (name, age_in_years, data)
         VALUES ({name: FBString}, {age: FBByte}, {data: FBString})", StatementName  );
```

which behind the scenes generates an instance of some type or value `StatementName` that could be re-used, and perhaps has it's own methods, and also stores the statement, for the user to use later.

This could be implemented in various ways, using closures, structs or a trait similar to `FromRow` or `Into<(&str, i8, &str)>` or some such. In some approaches it will be harder to make it so that the user can actually see the names of the parameters when they supply them. For this one could look at the implementation of `format!` to see how it does it

if implementing our own macro it is maybe even possible to recover the `:param` syntax in the above, but I think in this case it will be very challenging to correctly implement for all DSQL: it will be problematic if the user supplies a statement with `EXECUTE BLOCK`/`EXECUTE PROCEDURE` inside using that syntax which is using named parameters with the same style of syntax. We would then have to take care about only capturing free parameters.

in an approach involving structs that the user declares, which I think is what @fernandobatels  was proposing, it may be worthwhile to look at the `serde` crate for automatic derivation of serialization utilities.

**Comment by cytrinox** [comment]
isc_dsql_describe_bind() can handle named parameters directly, I've implemented this a few years ago in my own c++ database library https://github.com/informave/dbwtl/blob/master/src/dal/firebird/firebird_libfbclient.cc#L838 and it works pretty good in all cases.

**Comment by fernandobatels** [comment]
> isc_dsql_describe_bind() can handle named parameters directly, I've implemented this a few years ago in my own c++ database library https://github.com/informave/dbwtl/blob/master/src/dal/firebird/firebird_libfbclient.cc#L838 and it works pretty good in all cases.

I can't find any example of named parameters with isc_dsql_describe_bind in the firebird doc.

In your lib are declared some [bind methods](https://github.com/informave/dbwtl/blob/fed8f84ac42949d906a34b6a5a1b61b3a7b74356/include/dbwtl/dal/dal_interface.hh#L1389), but none using names instead of positions.  How, or where, you set the parameters name in this lib?

**Comment by fernandobatels** [comment]
@cytrinox we are finishing the #67 PR. In summary:

- Now we support named parameters using the original `:param` mark
- The named support are provided by a regex code, so we can't support complex query's
   - We are documenting this
- This new resource will be available in the next release(v0.12)
- The main point of out params is the `IntoParams` trait. You can implement this trait for whatever you need, also using the named parameters. 
    - Our lib provide the `IntoParams` derive to be used directly in structs
